fms/datasets/text.py

```diff
@@ -0,0 +1,79 @@
+import torch
+from torch.utils.data import Dataset
+import requests
+from fms.utils import tokenizers
+
+
+class FromString(Dataset):
+    """
+    Creates a dataset from a single text string and tokenizer.
+    Since all data comes from a single text, there are no bos/eos tokens used.
+    A pad token if specified, is used only on the final row. i.e.
+    `pad_token=None` is similar to drop_last in DataLoader.
+    """
+
+    def __init__(
+        self,
+        text: str,
+        tokenizer: tokenizers.BaseTokenizer,
+        seq_len: int = 1024,
+        pad_token: str = None,
+        device: torch.device = "cpu",
+        ignore_index: int = -100,
+    ):
+        tokens = tokenizer.tokenize(text)
+        ids = tokenizer.convert_tokens_to_ids(tokens)
+        self.ids = torch.tensor(ids, dtype=torch.long, device=device)
+        self.ignore_index = ignore_index
+        if pad_token is not None:
+            self.pad_id = tokenizer.convert_tokens_to_ids([pad_token])[0]
+        else:
+            self.pad_id = None
+        self.tokenizer = tokenizer
+        self.seq_len = seq_len
+
+    def to(self, device: torch.device):
+        self.ids = self.ids.to(device)
+        return self
+
+    def __getitem__(self, idx: int) -> torch.Tensor:
+        start_idx = idx * self.seq_len
+        end_idx = start_idx + self.seq_len + 1
+        if end_idx >= self.ids.shape[0]:
+            end_idx = self.ids.shape[0] - 1
+        input = self.ids[start_idx : end_idx - 1]
+        label = self.ids[start_idx + 1 : end_idx]
```

Comment:
labels here are just the sequence shifted by 1? For instance, if the entire text is:

```
a b c d e
```

and the seq_len=4

```
input = [a b c d]
labels = [b c d e]
```

Do we want this shift to be configurable?

Reply:
that is the expected shift for all causal LM tasks. We might want to add a parameter to control this for fine-tuning speculative decoding models, which is something we are about to start investigating for inference

Reply:
that is, if this is ever going to be used other than for testing

Reply:
speculative decoding sounds cool. if this is useful to use in testing of speculative decoding code, we can easily add it here later and the change would be backwards compatible.